syntax = "proto3";

package blockstorage;

service Raft
{
    rpc ReqVote(ReqVoteRequest) returns (ReqVoteReply) {}
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesReply) {}
    rpc AssertLeadership(AssertLeadershipRequest) returns (AssertLeadershipReply) {}
}

message ReqVoteRequest {
    uint32 term = 1;
    string candidateID = 2;
    uint32 lastLogIndex = 3;
    uint32 lastLogTerm = 4;
}

message ReqVoteReply {
    uint32 term = 1;
    bool voteGrantedFor = 2;
}

message AppendEntriesRequest {
    uint32 term = 1;
    string leaderID = 2;
    uint32 prevLogIndex = 3;
    uint32 prevLogTerm = 4;
    repeated LogEntry logEntry = 5;
}

message LogEntry {
    uint32 logIndex = 1;
    string key = 2;
    string value = 3;
}

message AppendEntriesReply {
    uint32 term = 1;
    bool success = 2;
}

message AssertLeadershipRequest {
    string leaderID = 1;
}

message AssertLeadershipReply {
    SystemState systemState = 1;
}

message SystemState {
    repeated FollowerID followerID = 1;
}

message FollowerID {
    string id = 1;
}